motionplan/check.go

```diff
@@ -0,0 +1,430 @@
+		poses[len(poses)-1],
+		startingInputs,
+		worldState,
+		nil, // no pb.Constraints
+		nil, // no plannOpts
+	); err != nil {
+		return err
+	}
+
+	// create a list of segments to iterate through
+	segments := make([]*ik.Segment, 0, len(poses)-wayPointIdx)
+
+	// iterate through remaining plan and append remaining segments to check
+	for i := wayPointIdx; i < len(offsetPlan.Path())-1; i++ {
+		segment, err := createSegment(sf, poses[i], poses[i+1], offsetPlan.Trajectory()[i], offsetPlan.Trajectory()[i+1])
+		if err != nil {
+			return err
+		}
+		segments = append(segments, segment)
+	}
+
+	// go through segments and check that we satisfy constraints
+	// TODO(RSDK-5007): If we can make interpolate a method on Frame the need to write this out will be lessened and we should be
+	// able to call CheckStateConstraintsAcrossSegment directly.
+	var totalTravelDistanceMM float64
+	for _, segment := range segments {
+		interpolatedConfigurations, err := interpolateSegment(segment, sfPlanner.planOpts.Resolution)
+		if err != nil {
+			return err
+		}
+		for _, interpConfig := range interpolatedConfigurations {
+			poseInPath, err := sf.Transform(interpConfig)
+			if err != nil {
+				return err
+			}
+
+			// Check if look ahead distance has been reached
+			currentTravelDistanceMM := totalTravelDistanceMM + poseInPath.Point().Distance(segment.StartPosition.Point())
+			if currentTravelDistanceMM > lookAheadDistanceMM {
+				return nil
+			}
+
+			// If we are working with a PTG plan the returned value for poseInPath will only
+			// tell us how far along the arc we have traveled. Since this is only the relative position,
+			//  i.e. relative to where the robot started executing the arc,
+			// we must compose poseInPath with segment.StartPosition to get the absolute position.
+			interpolatedState := &ik.State{Configuration: interpConfig, Frame: sf}
+
+			// Checks for collision along the interpolated route and returns a the first interpolated pose where a collision is detected.
+			if isValid, err := sfPlanner.planOpts.CheckStateConstraints(interpolatedState); !isValid {
+				return fmt.Errorf("found error between positions %v and %v: %s",
+					segment.StartPosition.Point(),
+					segment.EndPosition.Point(),
+					err,
+				)
+			}
+		}
+
+		// Update total traveled distance after segment has been checked
+		totalTravelDistanceMM += segment.EndPosition.Point().Distance(segment.StartPosition.Point())
+	}
+
+	return nil
+}
+
+// createSegment is a function to ease segment creation for solver frames.
+func createSegment(
+	sf *solverFrame,
+	currPose, nextPose spatialmath.Pose,
+	currInput, nextInput map[string][]referenceframe.Input,
+) (*ik.Segment, error) {
+	var currInputSlice, nextInputSlice []referenceframe.Input
+	var err error
+	if currInput != nil {
+		currInputSlice, err = sf.mapToSlice(currInput)
+		if err != nil {
+			return nil, err
+		}
+	}
+	nextInputSlice, err = sf.mapToSlice(nextInput)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
on `main` for this function we have
```
		// If we are working with a PTG plan we redefine the startConfiguration in terms of the endConfiguration.
		// This allows us the properly interpolate along the same arc family and sub-arc within that family.
		if relative {
			currInputSlice = nextInputSlice
		}
```
Why is this no longer needed?

Reply:
The logic for selecting what `currInputSlice` should be is now in the caller function rather than this one; it's simply passed through here.

Also, setting the start and end configurations identical is no longer the paradigm.